src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+            }
+            catch
+            {
+                chainPal.Dispose();
+                throw;
+            }
+
+            return chainPal;
+        }
+
+        private sealed class AndroidCertPath : IChainPal
+        {
+            public X509ChainElement[]? ChainElements { get; private set; }
+            public X509ChainStatus[]? ChainStatus { get; private set; }
+
+            public SafeX509ChainHandle? SafeHandle => null;
+
+            private SafeX509ChainContextHandle? _chainContext;
+            private bool _isValid;
+
+            public void Dispose()
+            {
+                if (_chainContext != null)
+                {
+                    _chainContext.Dispose();
+                }
+            }
+
+            public bool? Verify(X509VerificationFlags flags, out Exception? exception)
+            {
+                Debug.Assert(_chainContext != null);
+                exception = null;
+
+                if (!_isValid)
+                {
+                    // There is no way to bypass time, trusted root, name, or policy constraint
+                    // validation on Android. It will not build any chain without these all
+                    // being valid.
+                    X509VerificationFlags[] unsupportedFlags = new X509VerificationFlags[]
+                    {
+                        X509VerificationFlags.IgnoreNotTimeValid,
+                        X509VerificationFlags.AllowUnknownCertificateAuthority,
+                        X509VerificationFlags.IgnoreInvalidName,
+                        X509VerificationFlags.IgnoreInvalidPolicy,
+                    };
+                    foreach (X509VerificationFlags unsupported in unsupportedFlags)
+                    {
+                        if ((flags & unsupported) == unsupported)
+                        {
+                            exception = new PlatformNotSupportedException(SR.Format(SR.Chain_VerificationFlagNotSupported, unsupported));
+                            return default(bool?);
+                        }
+                    }
+
+                    return false;
+                }
+
+                return ChainVerifier.Verify(ChainElements!, flags);
+            }
+
+            internal void Initialize(
+                ICertificatePal cert,
+                X509Certificate2Collection? extraStore,
+                X509Certificate2Collection customTrustStore,
+                X509ChainTrustMode trustMode)
+            {
+                List<IntPtr> extraCerts = new List<IntPtr>() { cert.Handle };
+                if (extraStore != null)
+                {
+                    foreach (X509Certificate2 extraCert in extraStore)
+                    {
+                        extraCerts.Add(extraCert.Pal.Handle);
+                    }
+                }
+
+                List<IntPtr> customTrustCerts = new List<IntPtr>();
+                bool useCustomRootTrust = trustMode == X509ChainTrustMode.CustomRootTrust;
```

Comment:
There should be an assert here somewhere that if the mode isn't CustomRootTrust it's SystemTrust, so that the asserts help us track down where we need to support (or fail to support) when future modes are added.